xlsform_app/views.py

```diff
@@ -39,65 +56,98 @@ def handle_uploaded_file(f, temp_dir):
     destination.close()
     return xls_path
 
+def persist_file(persistence, dir_uuid, file_path):
+    persistence_dir = os.path.join(DJANGO_PERSISTENT_HOME, persistence.value, dir_uuid)
+    if not (os.access(persistence_dir, os.F_OK)):
+        os.makedirs(persistence_dir)
+    shutil.copy(file_path, persistence_dir)
 
-@xframe_options_exempt
-def index(request):
-    if request.method == 'POST':  # If the form has been submitted...
-        form = UploadFileForm(request.POST, request.FILES)  # A form bound to the POST data
-        if form.is_valid():  # All validation rules pass
 
-            error = None
-            warnings = None
+def convert_xlsform(file, persistence=Persistence.NONE):
+    error = None
+    warnings = None
 
-            filename, ext = os.path.splitext(request.FILES['file'].name)
+    filename, ext = os.path.splitext(file.name)
 
-            filename = clean_name(filename)
+    filename = clean_name(filename)
 
-            if not (os.access(DJANGO_TMP_HOME, os.F_OK)):
-                os.mkdir(DJANGO_TMP_HOME)
+    if not (os.access(DJANGO_TMP_HOME, os.F_OK)):
+        os.mkdir(DJANGO_TMP_HOME)
 
-            #Make a randomly generated directory to prevent name collisions
-            temp_dir = tempfile.mkdtemp(prefix=uuid.uuid4().hex, dir=DJANGO_TMP_HOME)
-            xml_path = os.path.join(temp_dir, filename + '.xml')
-            itemsets_url = None
+    #Make a randomly generated directory to prevent name collisions
+    dir_uuid = uuid.uuid4().hex
+    temp_dir = tempfile.mkdtemp(prefix=dir_uuid, dir=DJANGO_TMP_HOME)
+    xml_path = os.path.join(temp_dir, filename + '.xml')
+    relpath_itemsets_csv = None
 
-            relpath = os.path.relpath(xml_path, DJANGO_TMP_HOME)
+    relpath = os.path.relpath(xml_path, DJANGO_TMP_HOME)
 
-            #Init the output xml file.
-            fo = open(xml_path, "wb+")
-            fo.close()
-
-            try:
-                #TODO: use the file object directly
-                xls_path = handle_uploaded_file(request.FILES['file'], temp_dir)
-                warnings = []
-                json_survey = xls2json.parse_file_to_json(xls_path, warnings=warnings)
-                survey = pyxform.create_survey_element_from_dict(json_survey)
-                survey.print_xform_to_file(xml_path, warnings=warnings, pretty_print=False)
-
-                if has_external_choices(json_survey):
-                    # Create a csv for the external choices
-                    itemsets_csv = os.path.join(os.path.split(xls_path)[0],
-                                                "itemsets.csv")
+    #Init the output xml file.
+    fo = open(xml_path, "wb+")
+    fo.close()
+
+    try:
+        if ext == '.xml':
+            xml_path = handle_uploaded_file(file, temp_dir)
+            if(persistence != Persistence.NONE):
+                persist_file(persistence, dir_uuid, xml_path)
+            relpath = os.path.relpath(xml_path, DJANGO_TMP_HOME)
```

Comment:
Value of `xml_path` is mutated by `xml_path = handle_uploaded_file(file, temp_dir)`, it would be same though.

I have moved the initialization to `xml_path` and `relpath` to else block so both if and else block set their values accordingly.